lib/screens/settings/language.dart

```diff
@@ -50,27 +52,76 @@ class LanguagePage extends ConsumerWidget {
-        .compareTo(b.languageCode == 'en' ? 0 : 1));
+    final localizations = AppLocalizations.of(context)!;
+    final appLocale =
+        ref.watch(languageProvider) ?? Localizations.localeOf(context);
+    var supportedLocales = _getSupportedLocales(appLocale);
 
     return Scaffold(
-      appBar: AppBar(title: Text(loc.settingsLanguage)),
+      appBar: AppBar(title: Text(localizations.settingsLanguage)),
       body: ListView.separated(
-        itemCount: supported.length,
+        itemCount: supportedLocales.length,
         itemBuilder: (context, index) => ListTile(
-          title: Text(kLanguageNames[supported[index]] ??
-              supported[index].toLanguageTag()),
-          trailing: supported[index] == locale ? Icon(Icons.check) : null,
+          title: Text(kLanguageNames[supportedLocales[index]] ??
+              supportedLocales[index].toLanguageTag()),
+          trailing:
+              supportedLocales[index] == appLocale ? Icon(Icons.check) : null,
           onTap: () {
-            ref.read(languageProvider.notifier).set(supported[index]);
+            ref.read(languageProvider.notifier).set(supportedLocales[index]);
           },
         ),
         separatorBuilder: (context, index) => Divider(),
       ),
     );
   }
+
+  List<Locale> _getSupportedLocales(Locale appLocale) {
+    final supportedLocales = List.of(AppLocalizations.supportedLocales
+        .where((l) => !kSkipLocales.contains(l)));
+
+    // Move EN locales on top of the list
+    supportedLocales.sort((a, b) => (a.languageCode == 'en' ? 0 : 1)
+        .compareTo(b.languageCode == 'en' ? 0 : 1));
+
+    final deviceLocaleParts = Platform.localeName.split('.')[0].split('_');
+    var deviceLocales = <Locale>[];
+    switch (deviceLocaleParts.length) {
+      case 1:
+        deviceLocales.add(Locale(deviceLocaleParts[0]));
+        break;
+      case 2:
+        // We can have en_GB and en for example
+        deviceLocales.add(Locale(deviceLocaleParts[0], deviceLocaleParts[1]));
+        deviceLocales.add(Locale(deviceLocaleParts[0]));
+        break;
+      case 3:
+        // Locales with scriptCode has localeCode_scriptCode_countryCode
+        // structure. e.g. "zh_Hans_CN"
+        var languageCode = deviceLocaleParts[0];
+        var scriptCode = deviceLocaleParts[1];
+        deviceLocales.add(Locale.fromSubtags(
+            languageCode: languageCode, scriptCode: scriptCode));
+        break;
+    }
+
+    Locale? hintedLocale;
+    for (var deviceLocale in deviceLocales) {
+      if (!supportedLocales.contains(deviceLocale)) {
+        continue;
+      }
+
+      if (deviceLocale != appLocale) {
+        hintedLocale = deviceLocale;
+      }
+
+      // If device locale is supported but not the same as app locale we
+      // shouldn't check more general locale (only languageCode)
+      break;
+    }
+
+    if (hintedLocale != null && hintedLocale.languageCode != 'en')
```

Comment:
This part after the loop, and the `hintedLocale` variable in general looks like it can be assigned multiple times, while it can't. Could you please move this check and insert to the clause where `hintedLocale` is assigned, eliminating the need for extra variable and making this a bit easier to read?